### PR TITLE
add support for ipaddress to general name parsing

### PR DESCRIPTION
**Pull request by reaperhulk**
Update: PR only covers ip address now :)

### CODE REVIEW

File: src/cryptography/hazmat/backends/openssl/x509.py [hunk 1]
```diff
@@ -94,6 +96,39 @@ def _build_general_name(backend, gn):
...
         return x509.RegisteredID(x509.ObjectIdentifier(oid))
+    elif gn.type == backend._lib.GEN_DIRNAME:
+        return x509.DirectoryName(
+            _build_x509_name(backend, gn.d.directoryName)
+        )
+    elif gn.type == backend._lib.GEN_EMAIL:
+        data = backend._ffi.buffer(
+            gn.d.rfc822Name.data, gn.d.rfc822Name.length
+        )[:].decode("ascii")
+        parsed = parseaddr(data)
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/x509.py]: Please use tuple unpacking here, rather than `[1]`.

Also, is it really correct to handle `<>` formatted addresses here? I've only ever seen an RFC822Name with a "normal" email addr.

Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/x509.py]: I couldn't decide. The proper way to encode an rfc822name is just the bare email address, but there's nothing to stop you from doing something far more horrible. Maybe we should just say we aren't going to IDNA decode anything but the standard representation though?

Comment (alex) [inline src/cryptography/hazmat/backends/openssl/x509.py]: Maybe we should just parse error on anything else, absent actual usage in
the real world?

On Tue, Apr 28, 2015 at 7:03 PM, Paul Kehrer <notifications@github.com>
wrote:

> In src/cryptography/hazmat/backends/openssl/x509.py
> <https://github.com/pyca/cryptography/pull/1872#discussion_r29297505>:
>
> > @@ -94,6 +96,39 @@ def _build_general_name(backend, gn):
> >      elif gn.type == backend._lib.GEN_RID:
> >          oid = _obj2txt(backend, gn.d.registeredID)
> >          return x509.RegisteredID(x509.ObjectIdentifier(oid))
> > +    elif gn.type == backend._lib.GEN_DIRNAME:
> > +        return x509.DirectoryName(
> > +            _build_x509_name(backend, gn.d.directoryName)
> > +        )
> > +    elif gn.type == backend._lib.GEN_EMAIL:
> > +        data = backend._ffi.buffer(
> > +            gn.d.rfc822Name.data, gn.d.rfc822Name.length
> > +        )[:].decode("ascii")
> > +        parsed = parseaddr(data)
>
> I couldn't decide. The proper way to encode an rfc822name is just the bare
> email address, but there's nothing to stop you from doing something far
> more horrible. Maybe we should just say we aren't going to IDNA decode
> anything but the standard representation though?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1872/files#r29297505>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

File: src/cryptography/hazmat/backends/openssl/x509.py [hunk 2]
```diff
@@ -94,6 +96,39 @@ def _build_general_name(backend, gn):
     elif gn.type == backend._lib.GEN_RID:
         oid = _obj2txt(backend, gn.d.registeredID)
         return x509.RegisteredID(x509.ObjectIdentifier(oid))
+    elif gn.type == backend._lib.GEN_DIRNAME:
+        return x509.DirectoryName(
+            _build_x509_name(backend, gn.d.directoryName)
+        )
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/x509.py]: This couldn't go into its own PR?

File: src/cryptography/hazmat/backends/openssl/x509.py [hunk 3]
```diff
@@ -94,6 +96,39 @@ def _build_general_name(backend, gn):
...
+        else:
+            return x509.RFC822Name(data)
+    elif gn.type == backend._lib.GEN_IPADD:
+        return x509.IPAddress(
+            ipaddress.ip_address(
+                backend._ffi.buffer(
+                    gn.d.iPAddress.data, gn.d.iPAddress.length
+                )[:]
+            )
+        )
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/x509.py]: Same here, these are both much easier to review than the RFC bit

**Comment by alex** [comment]
I'd really prefer if we could land "everything but email" first, that stuff is MUCH easier to review :-)

**Comment by reaperhulk** [comment]
retest this please
